### PR TITLE
Remove dependency on Process.clock_getres

### PR DESCRIPTION
**Pull request by copiousfreetime**

### CODE REVIEW

**Comment by copiousfreetime** [comment]
Closes #76

**Comment by copiousfreetime** [comment]
Any opinions on the [somewhat hacky mechanism](https://github.com/copiousfreetime/hitimes/blob/3461eff15a430ab50ef513cef043d5410ce8d2c9/lib/hitimes/initialize.rb#L30-L61) to determine best resolution clock @eregon?

**Comment by eregon** [comment]
That sounds a bit better, but still rather fragile.
It also sounds like a non-trivial overhead on the first usage.
I think being a non-deterministic is a big issue here, for instance it could lead to using one clock for one process run, and another for the next run on the same machine.
That's problematic for benchmarking, especially if one is CLOCK_MONOTONIC_RAW and the other CLOCK_MONOTONIC, because they don't have the same definition of a second.

IMHO hardcoding per platform would be the best here, the worst that could happen is to get a slightly less precise clock on an uncommon OS, IMHO not a big deal and easy enough to fix.
So I think starting with just `macos ? CLOCK_MONOTONIC_RAW : CLOCK_MONOTONIC` would be the best.

Or simplify even further and just always use `CLOCK_MONOTONIC`, because I believe that's the clock everyone should use for benchmarking and accurate time measurements.
`CLOCK_MONOTONIC_RAW` doesn't actually represent seconds and so could be quite off from a real second in some rare but possible cases (IIRC, because the Linux man page doesn't make this very explicit).

**Comment by copiousfreetime** [comment]
Thanks for the feedback @eregon I'm not feeling particularly good about the 'pick the clock on require' approach. I think I'm going to do some testing on as many platforms as I can find and see what comes out. I'll probably go with the hardcode-by-platform approach, with fallback.

**Comment by eregon** [comment]
Yes, I think that ultimately that is the best solution, even though it's clearly not the most convenient.
Also the "worst case" is a slightly less precise clock, which seems not too bad, compared to other approaches.

I think CLOCK_MONOTONIC is even precise to microseconds on most platforms, and that's probably more than good enough. I would think extremely few use cases actually need nanosecond precision.

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -13,18 +13,21 @@ blocks:
...
-                - 3.0.4
-                - 3.1.2
-                - jruby-9.3.4.0
-                - truffleruby-22.0.0.2
+                - 2.7.7
+                - 3.0.5
+                - 3.1.3
+                - 3.2.0
+                - jruby-9.3.9.0
+                - jruby-9.4.0.0
```
Comment (eregon) [inline .semaphore/semaphore.yml:26]: If you want to use the latest release of each release series you could use `2.6, 2.7, '3.0', 3.1, 3.2, jruby-9.3, jruby-9.4, truffleruby-22, truffleruby+graalvm-22

Comment (eregon) [reply .semaphore/semaphore.yml:26]: `truffleruby-22, truffleruby+graalvm-22` are redundant (they will both test in `--native` mode, the default), so I'd keep just `truffleruby-22`

Comment (eregon) [reply .semaphore/semaphore.yml:26]: Or even just `truffleruby` (i.e., the latest release) so then it always tests the latest release of TruffleRuby, which is always supported

Comment (eregon) [reply .semaphore/semaphore.yml:26]: AH, although I have no idea what syntax SemaphoreCI supports :sweat_smile: (I thought it was GitHub Actions)

Comment (copiousfreetime) [reply .semaphore/semaphore.yml:26]: Its rbenv under the covers - so I'll play around with it.

Comment (copiousfreetime) [reply .semaphore/semaphore.yml:26]: @eregon Its rbenv && ruby-build under the covers (which I didn't know you helped maintain - thanks btw!) - and doing just `major.minor` with that doesn't appear to be supported.  github actions must be doing some `.rbenv/versions/` symlinking under the covers.

Is doing `rbenv install major.minor` a feature request I should drop into rbenv/ruby-build ?

```
semaphore@semaphore-vm:~$ rbenv install --version
ruby-build 20221225-9-g697bcff

semaphore@semaphore-vm:~$ rbenv install 3.1
ruby-build: definition not found: 3.1

The following versions contain `3.1' in the name:
  2.3.1
  3.1.0-dev
  3.1.0-preview1
  3.1.0
  3.1.1
  3.1.2
...
```

Comment (eregon) [reply .semaphore/semaphore.yml:27]: I would tend to think such resolution should happen before it hits ruby-build/rbenv. For instance https://github.com/ruby/setup-ruby does that on its own. Also it sounds pretty slow for a CI if it has to compile CRuby from source :sweat_smile: (but probably it does some caching?)

But maybe it makes sense for ruby-build too, feel free to file a suggestion there.

Comment (copiousfreetime) [reply .semaphore/semaphore.yml:27]: Semaphore does that too with their `sem-version` tool - which is roughly equivalent to the github action, its what is used in the pipeline to pick the ruby version, so it uses a prebuilt one if it has one, and then delegates to ruby-build if it doesn't have it pre-built.

Comment (copiousfreetime) [reply .semaphore/semaphore.yml:27]: I've opened a ticket with semaphore it looks like they just haven't `sem-version` to handle 3.2 patch versions yet. I'll update all of this when/if that is resolved.

**Comment by Freaky** [comment]
FWIW for [monotime](https://github.com/Freaky/monotime) I just went with `CLOCK_UPTIME_RAW` for macOS and `CLOCK_MONOTONIC` for everything else.

I only *just* justified the former on the basis of:

1. It benchmarked faster on two different architectures, albeit only 10-25%
2. It's higher resolution, which I don't think matters much for Ruby but, hey
3. Rust did it too, and that's what I'm mostly emulating anyway :P

As for your attempt to detect resolution by counting zero bits, I tried that myself out of interest -- here's `CLOCK_MONOTONIC_FAST`, 1,000 runs of majority-of 10,000 samples:

```
=> {"1ns"=>813, "2ns"=>143, "4ns"=>32, "8ns"=>11, "16ns"=>1}
```

And sure - it has 1ns resolution, but here's what you get if you bucket 10,000 back-to-back calls:

```
{562332641125037 => 83,
 562332642124615 => 704,
 562332643124423 => 704,
 562332644124251 => 735,
 562332645124119 => 743,
 562332646124088 => 824,
 562332647124126 => 749,
 562332648124025 => 736,
 562332649123923 => 734,
 562332650124002 => 772,
 562332651123950 => 789,
 562332652123989 => 741,
 562332653124108 => 746,
 562332654123996 => 821,
 562332655123944 => 119}
```

It only updates about once per millisecond.  In contrast 10,000 back-to-back `CLOCK_MONOTONIC` calls gives me 10,000 unique times, an average of 390ns apart.

Also, this:

```ruby
# Sort them by the resolution - we want the smallest one first
      ids_and_resolutions.sort_by! { |pair| pair[1] }
      best_clock_and_resolution = ids_and_resolutions[0]
```

To quote Ruby docs: "For duplicates returned by the block, the ordering is indeterminate, and may be unstable".

If you had multiple clocks with 1ns resolution you can end up selecting a different one each time.  If you were to keep this approach it might be wise to include something like an array index (with the options sorted by general preference) to act as a tie-breaker.  [Like so](https://github.com/Freaky/monotime/blob/2248eeebf1cc4f396f72811b41b175ece726aa22/lib/monotime/instant.rb#L84-L97).

I decided against it.

**Comment by copiousfreetime** [comment]
Well - finally got back to this after way too long.

I did a fair bit of testing and I'm going for a test of `CLOCK_MONOTONIC_RAW`, then `CLOCK_MONOTINIC` and final fall back to `CLOCK_REALTIME`

Doing the same as @Freaky and dropping a warning to the user if we fallback to `CLOCK_REALTIME`, and a super failsafe that if none of the 3 are found, its an exception. Not that I expect that to actually happen, but I've had bugs show up that I didn't expect before. I decided to not use `CLOCK_UPTIME_RAW` as it stops timing during if the computer goes to sleep. At least according to the documentation.

Thank you @eregon and @Freaky for you input!

**Comment by copiousfreetime** [comment]
For reference, the clock determination logic is now:
https://github.com/copiousfreetime/hitimes/blob/8b467d2c1602167973ef6ea8d018ff1c6932e28e/lib/hitimes/initialize.rb#L74-L99

**Comment by eregon** [comment]
Looks good!

**Pull request merged**